File: unleash/src/main/java/com/silvercar/unleash/strategy/FlexibleRolloutStrategy.java
package com.silvercar.unleash.strategy;

import java.util.Map;
import com.annimon.stream.Optional;

import com.annimon.stream.function.Supplier;
import com.silvercar.unleash.UnleashContext;

public class FlexibleRolloutStrategy implements Strategy {
    protected static final String PERCENTAGE = "rollout";
    protected static final String GROUP_ID = "groupId";

    private Supplier<String> randomGenerator;

    public FlexibleRolloutStrategy() {
        this.randomGenerator = () ->  Math.random() * 100 + "";
    }

    public FlexibleRolloutStrategy(Supplier<String> randomGenerator) {
        this.randomGenerator = randomGenerator;
    }

    @Override
    public String getName() {
        return "flexibleRollout";
    }

    @Override
    public boolean isEnabled(Map<String, String> parameters) {
        return false;
    }

    private String resolveStickiness(String stickiness, UnleashContext context) {
        switch (stickiness) {
            case "userId": return context.getUserId();
            case "sessionId": return context.getSessionId();
            case "random": return randomGenerator.get();
            default:
                String userId = context.getUserId();
                if (userId != null && !userId.isEmpty()) return userId;

                String sessionId = context.getSessionId();
                if (sessionId != null && !sessionId.isEmpty()) return sessionId;

                return this.randomGenerator.get();
        }
    }

    @Override
    public boolean isEnabled(Map<String, String> parameters, UnleashContext unleashContext) {
        final String stickiness = getStickiness(parameters);
        final String stickinessId = resolveStickiness(stickiness, unleashContext);
        final int percentage = StrategyUtils.getPercentage(parameters.get(PERCENTAGE));
        final String groupId = Optional.ofNullable(parameters.get(GROUP_ID)).orElse("");

        if (stickinessId != null && !stickinessId.isEmpty()) {
            final int normalizedUserId = StrategyUtils.getNormalizedNumber(stickinessId, groupId);
            return percentage > 0 && normalizedUserId <= percentage;
        } else {
            return false;
        }
    }

    private String getStickiness(Map<String, String> parameters) {
        Optional<String> stickiness = Optional.ofNullable(parameters.get("stickiness"));
        return stickiness.orElse("default");
    }
}

File: unleash/src/main/java/com/silvercar/unleash/strategy/FlexibleRolloutStrategy.kt
package com.silvercar.unleash.strategy

import com.silvercar.unleash.UnleashContext
import com.silvercar.unleash.strategy.Strategy.Companion.GROUP_ID
import com.silvercar.unleash.strategy.Strategy.Companion.ROLLOUT

class FlexibleRolloutStrategy(private val randomGenerator: RandomGenerator) :
  Strategy {
  private val strategyUtils: StrategyUtils = StrategyUtils()

  override val name: String
    get() = STRATEGY_NAME

  override fun isEnabled(parameters: Map<String, String>): Boolean {
    return false
  }

  private fun resolveStickiness(stickiness: String, context: UnleashContext): String {
    return when (stickiness) {
      "userId" -> context.userId
      "sessionId" -> context.sessionId
      "random" -> randomGenerator.get()
      else -> if (context.userId.isNotEmpty()) {
        context.userId
      } else if (context.sessionId.isNotEmpty()) {
        context.sessionId
      } else {
        randomGenerator.get()
      }
    }
  }

  override fun isEnabled(
    parameters: Map<String, String>,
    unleashContext: UnleashContext
  ): Boolean {
    val stickiness = getStickiness(parameters)
    val stickinessId = resolveStickiness(stickiness, unleashContext)
    val percentage = strategyUtils.getPercentage(parameters[ROLLOUT])
    val groupId = parameters[GROUP_ID] ?: ":"

    return if (stickinessId.isNotEmpty()) {
      val normalizedUserId = strategyUtils.getNormalizedNumber(stickinessId, groupId)
      percentage > 0 && normalizedUserId <= percentage
    } else {
      false
    }
  }

  private fun getStickiness(parameters: Map<String, String>): String {
    val stickiness = parameters[STICKINESS] ?: ""
    return if (stickiness.isEmpty()) "default" else stickiness
  }

  companion object {
    private const val STRATEGY_NAME = "flexibleRollout"
    const val STICKINESS = "stickiness"
  }
}
